File: app/src/main/java/net/sarasarasa/basemvp/base/MvpActivity.java
package net.sarasarasa.basemvp.base;

import android.os.Bundle;
import android.widget.Toast;

import androidx.annotation.LayoutRes;
import androidx.appcompat.app.AppCompatActivity;

@SuppressWarnings("unchecked")
public abstract class MvpActivity<P extends IMvpPresenter> extends AppCompatActivity implements IMvpView {

    protected P mPresenter;
    protected boolean isRetainInstance = false;

    @LayoutRes
    protected int layoutResID;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(layoutResID);

        mPresenter = createPresenter();
        if (mPresenter == null) {
            throw new NullPointerException("Presenter is null");
        }
        mPresenter.attachView(this);


        bindView();
        initView();
    }


    @Override
    protected void onStart() {
        super.onStart();

    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mPresenter.detachView(isRetainInstance);
    }

    /**
     * Instantiate a mPresenter instance
     *
     * @return The {@link MvpPresenter} for this view
     */
    protected abstract P createPresenter();

    protected abstract void initView();

    protected abstract void bindView();

    @Override
    public void showMessage(String message) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show();
    }
}

File: kotlin/src/main/java/net/sarasarasa/basemvpkotlin/base/MvpActivity.kt
package net.sarasarasa.basemvpkotlin.base

import android.os.Bundle
import android.widget.Toast
import androidx.annotation.LayoutRes

import androidx.appcompat.app.AppCompatActivity

@Suppress("UNCHECKED_CAST")
abstract class MvpActivity<in V : IMvpView, P : IMvpPresenter<V>> : AppCompatActivity(), IMvpView {

    protected var mPresenter: P? = null
    protected var isRetainInstance = false
    @LayoutRes
    protected var layoutResID: Int? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        layoutResID?.let { setContentView(it) }

        mPresenter = createPresenter()
        if (mPresenter == null) {
            throw NullPointerException("Presenter is null")
        }
        mPresenter?.attachView(this as V)
        initView()
        bindView()
    }

    override fun onDestroy() {
        super.onDestroy()
        mPresenter?.detachView(isRetainInstance)
    }

    /**
     * Instantiate a mPresenter instance
     *
     * @return The [MvpPresenter] for this view
     */
    protected abstract fun createPresenter(): P

    protected abstract fun initView()
    protected abstract fun bindView()

    override fun showMessage(message: String) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show();
    }
}